PL2/PL2-parser/src/main/java/nl/tudelft/pl2016gr2/parser/controller/FullGfaReader.java

```diff
@@ -3,153 +3,218 @@
+   *
+   * @param node     the node to add the bases to.
+   * @param chars    the character array of the node line in the GFA file.
+   * @param curIndex the current index; where the bases start in the line.
+   * @return the new index, after reading the genomes.
+   */
+  private int parseNodeBases(Node node, char[] chars, int curIndex) {
+    int index = curIndex;
+    index = skipTillCharacter(chars, index, '\t', 1);
+    node.setBases(new String(chars, curIndex, index - curIndex));
+    return index;
   }
 
   /**
-   * Private method for handling an "S" as first character on the line of a gfa file.
-   * @param sc The scanner which is used to parse the file.
+   * Parse the genomes of a node.
+   *
+   * @param node     the node to add the genomes to.
+   * @param chars    the character array of the node line in the GFA file.
+   * @param curIndex the current index; where the bases end in the line.
+   * @return the new index, after reading the genomes.
    */
-  private void handleS(Scanner sc) {
-    int nodeId = sc.nextInt();
-    String bases = sc.next();
-    String line = sc.nextLine();
-    String[] words = line.split("\\s+");
-    ArrayList<String> nodeGens = extractGenomes(words[2]);
-    int orientation = extractStart(words[6]);
-    //words[3] = CRD:Z:XXXX
-    //words[4] = CRDCTG:Z:XXXX
-    //words[5] = CTG:Z:XXXX;XXXX;...
-    //words[6] = START:Z:INT
-    Node node = originalGraph.getNode(nodeId);
+  private int parseNodegenomes(Node node, char[] chars, int curIndex) {
+    int index = curIndex;
+    index = skipTillCharacter(chars, index, ':', 2);
+    ++index;
+    int startIndex = index;
+    ArrayList<String> nodeGens = new ArrayList<>();
+    while (chars[index] != '\t') {
+      ++index;
+      while (chars[index] != ';' && chars[index] != '\t') {
+        ++index;
+      }
+      nodeGens.add(new String(chars, startIndex, index - startIndex));
+      startIndex = index + 1;
+    }
     node.setGenomes(nodeGens);
-    node.setBases(bases);
-    node.setSequenceLength(bases.length());
+    return index;
+  }
+
+  /**
+   * Parse the orientation of a node.
+   *
+   * @param node     the node of which the orientation is read.
+   * @param chars    the character array of the node line in the GFA file.
+   * @param curIndex the current index; where the genomes end in the line.
+   */
+  private void parseNodeOrientation(Node node, char[] chars, int curIndex) {
+    int index = skipTillCharacter(chars, curIndex, '\t', 3);
+    index = skipTillCharacter(chars, index, ':', 2) + 1;
+    int orientation = 0;
+    while (index < chars.length) {
+      orientation *= SHIFT_BY_BASE_10;
+      orientation += chars[index++] - '0';
+    }
     node.setAlignment(orientation);
   }
 
   /**
-   * Private method for handling an "L" as first character on the line of a gfa file.
-   * @param sc The scanner which is used to parse the file.
+   * Parse the header part of the GFA file, which contains all of the genomes.
```

Comment:
This is correct usage of 'which'. the adjective clause 'which contains all of the genomes' is only informative: leaving it out retains the correct meaning of the sentence.
:)